### PR TITLE
Issue 8 - Add support for markdown heading h1-h6, p

### LINKED ISSUE #8: Add Markdown support for heading h1-h6, and p
- [x]  Add support for .md
- [x]  Add checking for file .txt and .md
- [x]  Add function to convert markdown heading h1-h6, and p to HTML
- [x]  Update readme.md

### PR DESCRIPTION
**Pull request by Kevan-Y**
Fixes #8 
- [x]  Add support for .md
- [x]  Add checking for file .txt and .md
- [x]  Add function to convert markdown heading h1-h6, and p to HTML
- [x] Update readme.md

This function below is to process markdown. Will be able this function to add more markdown in the future
https://github.com/Kevan-Y/OSD_SSG/blob/05a3f8d6e5a47e43e83ec5e20b360106dda0ac3b/src/index.ts#L128-L134

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -2,7 +2,7 @@
 
 **OSD_SSG** is a static site generator, implemented as part of Seneca College's OSD600 course.
 
-> An *SSG* is a tool for generating a complete HTML web site from raw data and files, without having to author any HTML by hand. Imagine that you have a folder of text files that you want to turn into a website. An SSG allows you to enter a simple command that creates .html output files from a given set of input files.
+> An _SSG_ is a tool for generating a complete HTML web site from raw data and files, without having to author any HTML by hand. Imagine that you have a folder of text files that you want to turn into a website. An SSG allows you to enter a simple command that creates .html output files from a given set of input files.
```
Comment (DukeManh) [inline README.md]: Aren't these equivalent to each other?

File: src/index.ts [hunk 2]
```diff
@@ -78,6 +78,61 @@ if (stylesheet) {
...
 }
 
+/**
+ * Process markdown for H1 (e.g: # headhing 1)
+ * @param data Data to be processed
+ * @return return array where h1 markdown is processed
+ */
+const processH1Markdown = (data: string[]): string[] => {
+  return data.map((text) => {
+    if (text.indexOf('# ') !== -1) {
```
Comment (DukeManh) [inline src/index.ts]: This check is redundant. The next line already checks if `# ` exists.

File: src/index.ts [hunk 3]
```diff
@@ -78,6 +78,61 @@ if (stylesheet) {
...
+};
+
+/**
+ * Process markdown for H2 (e.g: ## headhing 2)
+ * @param data Data to be processed
+ * @return return array where h2 markdown is processed
+ */
+const processH2Markdown = (data: string[]): string[] => {
+  return data.map((text) => {
+    if (text.indexOf('## ') !== -1) {
```
Comment (DukeManh) [inline src/index.ts]: Same here.

File: src/index.ts [hunk 4]
```diff
@@ -78,6 +78,61 @@ if (stylesheet) {
...
+};
+
+/**
+ * Process markdown for P
+ * @param data Data to be processed
+ * @return return array where p markdown is processed
+ */
+const processPMarkdown = (data: string[]): string[] => {
+  return data.map((text) => {
+    if (text.substr(0, 1) !== '<' && text.substr(text.length - 1, 1) !== '>')
```
Comment (DukeManh) [inline src/index.ts]: What's the purpose of checking if `text` starts with `<` ends with `>`?

Comment (DukeManh) [reply src/index.ts]: Small tip, `text.substr(text.length - 1, 1)` is the same as `text.substr(text.length - 1)` ( without the second argument).

File: src/index.ts [hunk 5]
```diff
@@ -110,7 +165,7 @@ const processFile = (filePath: string, isIndex: boolean): string => {
                 <link rel="stylesheet" href="${
                   path.relative(path.dirname(filePath), input) || './'
                 }/index.css"> 
-                <title>${title || path.basename(filePath, '.txt')}</title>`;
+                <title>${title || path.basename(filePath, isMd ? '.md' : '.txt')}</title>`;
```
Comment (DukeManh) [inline src/index.ts]: You can write `path.basename(filePath, path.extname)` to remove the extension from name.

File: src/index.ts [hunk 6]
```diff
@@ -156,7 +218,10 @@ const generateHTMLs = (pathName: string): string[] => {
         const markup = processFile(filePath, false);
         if (markup) {
           const relativeFolder = relative ? `/${path.relative(input, path.dirname(filePath))}` : '';
-          const dist = `${output}${relativeFolder}/${path.basename(filePath, '.txt')}.html`;
+          const dist = `${output}${relativeFolder}/${path.basename(
+            filePath,
+            path.extname(filePath).toLocaleLowerCase() === '.txt' ? '.txt' : '.md'
```
Comment (DukeManh) [inline src/index.ts]: Same here, `path.basename(filepath, path.extname(filePath))`

File: src/index.ts [hunk 7]
```diff
@@ -196,10 +261,17 @@ try {
...
   if (!markup) {
-    logError('Input file must be .txt');
+    logError('Input file must be .txt or .md');
   }
 
-  fs.writeFileSync(`${output}/${path.basename(input, '.txt')}.html`, markup, { flag: 'w' });
+  fs.writeFileSync(
+    `${output}/${path.basename(
+      input,
+      path.extname(input).toLocaleLowerCase() === '.txt' ? '.txt' : '.md'
```
Comment (DukeManh) [inline src/index.ts]: Same here.

File: src/index.ts [hunk 8]
```diff
@@ -86,17 +141,17 @@ if (stylesheet) {
...
-  if (extension !== '.txt') {
+  if (extension !== '.txt' && extension !== '.md') {
     return '';
   }
 
+  const isMd = extension === '.md';
   const text = fs.readFileSync(filePath, 'utf-8');
 
   // title is before the first 2 blank lines of the text
   const titleAndContent = text.split(/\n\n\n/);
```
Comment (DukeManh) [inline src/index.ts:136]: The title should only be parsed like this when the input file is `.txt`. Heading in markdown doesn't have to be before the first two blank lines. I'd say you can skip parsing the title if the file is `.md`.

**Review by DukeManh — CHANGES_REQUESTED** [review]

File: src/index.ts [hunk 9]
```diff
@@ -78,6 +78,61 @@ if (stylesheet) {
...
+ * Process markdown for H1 (e.g: # headhing 1)
+ * @param data Data to be processed
+ * @return return array where h1 markdown is processed
+ */
+const processH1Markdown = (data: string[]): string[] => {
+  return data.map((text) => {
+    if (text.indexOf('# ') !== -1) {
+      if (text.substr(0, 2) === '# ') return `<h1>${text.substr(2)}</h1>`;
+    }
+    return `${text}`;
```
Comment (DukeManh) [inline src/index.ts]: Return `text` directly as text is already a string.

File: README.md (hunk 1 above)
Comment (Kevan-Y) [reply README.md]: Oh, I can undo it. My autoformatting in my VS code change it.

File: src/index.ts (hunk 4 above)
Comment (Kevan-Y) [reply src/index.ts]: This is checking if it is wrapped to something like h1. If it is, do not need to turn into paragraph

File: src/index.ts (hunk 5 above)
Comment (Kevan-Y) [reply src/index.ts]: changed to `path.parse(filePath).name`

File: src/index.ts (hunk 6 above)
Comment (Kevan-Y) [reply src/index.ts]: Updated

File: src/index.ts (hunk 7 above)
Comment (Kevan-Y) [reply src/index.ts]: Updated

File: src/index.ts (hunk 8 above)
Comment (Kevan-Y) [reply src/index.ts:136]: On my side it doesn't affect anything. Since this line of code has an issue.

File: src/index.ts (hunk 9 above)
Comment (Kevan-Y) [reply src/index.ts]: removed

File: src/index.ts [hunk 10]
```diff
@@ -78,6 +78,54 @@ if (stylesheet) {
   fs.copyFileSync('src/styles/index.css', `${output}/index.css`);
 }
 
+/**
+ * Process markdown for H1 (e.g: # headhing 1)
```
Comment (DukeManh) [inline src/index.ts]: Can you change the spelling to `heading`?

File: src/index.ts [hunk 11]
```diff
@@ -78,6 +78,54 @@ if (stylesheet) {
...
+ * @return return array where h1 markdown is processed
+ */
+const processH1Markdown = (data: string[]): string[] => {
+  return data.map((text) => {
+    return text.replace(/^# (.*$)/gi, '<h1>$1</h1>');
+  });
+};
+
+/**
+ * Process markdown for H2 (e.g: ## headhing 2)
```
Comment (DukeManh) [inline src/index.ts]: same here

File: src/index.ts [hunk 12]
```diff
@@ -78,6 +78,54 @@ if (stylesheet) {
...
+ * @param data Data to be processed
+ * @return return array where h2 markdown is processed
+ */
+const processH2Markdown = (data: string[]): string[] => {
```
Comment (DukeManh) [inline src/index.ts]: Can you combine `processH1` and `processH2` into 1 single function and process all headings from `h1` to `h6`?
```
text.replace(/^\s*(#{1,6})\s+(.+)$/gm, (match, hash, title) => {
					const tag = `h${hash.length}`;
					return `<${tag}>${title}</${tag}>`;
				});
```

**Review by DukeManh — CHANGES_REQUESTED** [review]

File: src/index.ts (hunk 10 above)
Comment (Kevan-Y) [reply src/index.ts]: sure

File: src/index.ts (hunk 11 above)
Comment (Kevan-Y) [reply src/index.ts]: sure

File: src/index.ts (hunk 12 above)
Comment (Kevan-Y) [reply src/index.ts]: changed

**Pull request merged**